op-supervisor/supervisor/backend/db/logs/db.go

```diff
@@ -206,31 +199,63 @@ func (db *DB) Get(blockNum uint64, logIdx uint32) (common.Hash, error) {
 // If the log is determined to conflict with the canonical chain, then ErrConflict is returned.
 // logIdx is the index of the log in the array of all logs in the block.
 // This can be used to check the validity of cross-chain interop events.
-func (db *DB) Contains(blockNum uint64, logIdx uint32, logHash common.Hash) (entrydb.EntryIdx, error) {
+// The block-seal of the blockNum block, that the log was included in, is returned.
+// This seal may be fully zeroed, without error, if the block isn't fully known yet.
+func (db *DB) Contains(blockNum uint64, logIdx uint32, logHash common.Hash) (types.BlockSeal, error) {
 	db.rwLock.RLock()
 	defer db.rwLock.RUnlock()
 	db.log.Trace("Checking for log", "blockNum", blockNum, "logIdx", logIdx, "hash", logHash)
 
 	evtHash, iter, err := db.findLogInfo(blockNum, logIdx)
 	if err != nil {
-		return 0, err // may be ErrConflict if the block does not have as many logs
+		return types.BlockSeal{}, err // may be ErrConflict if the block does not have as many logs
 	}
 	db.log.Trace("Found initiatingEvent", "blockNum", blockNum, "logIdx", logIdx, "hash", evtHash)
 	// Found the requested block and log index, check if the hash matches
 	if evtHash != logHash {
-		return 0, fmt.Errorf("payload hash mismatch: expected %s, got %s", logHash, evtHash)
+		return types.BlockSeal{}, fmt.Errorf("payload hash mismatch: expected %s, got %s", logHash, evtHash)
 	}
-	return iter.NextIndex(), nil
+	// Now find the block seal after the log, to identify where the log was included in.
+	err = iter.TraverseConditional(func(state IteratorState) error {
+		_, n, ok := state.SealedBlock()
+		if !ok { // incomplete block data
+			return nil
+		}
+		if n == blockNum {
+			return entrydb.ErrStop
+		}
+		if n > blockNum {
+			return entrydb.ErrDataCorruption
+		}
+		return nil
+	})
+	if err == nil {
+		panic("expected iterator to stop with error")
```

Comment:
Why panic here instead of returning an error? We can always handle the error by shutting things down if that's what we really want.

Reply:
Perhaps a graceful shutdown would be good in the future , but at the moment `panic` is the right call because we don't expect it to be possible in the first place -- traverse conditional proceeds until an error is encountered (either an error in the database, or an error due to the conditional failing).

If `TraverseConditional` ever returned without error, we'd be in totally undefined territory, fit for a panic. As for development-operations, we would want to immediately address the underlying issue if one should exist, rather than implement graceful exit, because we fully rely on `TraverseConditional` behaving in this way.